src/Editor/GUI/Editors/CurveEditor.cpp

```diff
@@ -0,0 +1,744 @@
+                                if(p < 0.5) {
+                                double f = 2 * p;
+                                return 0.5 * (f * f * f - f * sin(f * pi));
+                                }
+                                else {
+                                double f = (1 - (2*p - 1));
+                                return 0.5 * (1 - (f * f * f - f * sin(f * pi))) + 0.5;
+                                } */
+            double s = 1.70158f * 1.525f;
+            if (p < 0.5) {
+                return p *= 2, 0.5 * p * p * (p*s + p - s);
+            } else {
+                return p = p * 2 - 2, 0.5 * (2 + p*p*(p*s + p + s));
+            }
+        }
+
+#           define tween$bounceout(p) ( \
+                (p) < 4/11.0 ? (121 * (p) * (p))/16.0 : \
+                (p) < 8/11.0 ? (363/40.0 * (p) * (p)) - (99/10.0 * (p)) + 17/5.0 : \
+                (p) < 9/10.0 ? (4356/361.0 * (p) * (p)) - (35442/1805.0 * (p)) + 16061/1805.0 \
+                           : (54/5.0 * (p) * (p)) - (513/25.0 * (p)) + 268/25.0 )
+
+        case TYPE::BOUNCEIN: {
+            return 1 - tween$bounceout(1 - p);
+        }
+
+        case TYPE::BOUNCEOUT: {
+            return tween$bounceout(p);
+        }
+
+        case TYPE::BOUNCEINOUT: {
+            if (p < 0.5) {
+                return 0.5 * (1 - tween$bounceout(1 - p * 2));
+            } else {
+                return 0.5 * tween$bounceout((p * 2 - 1)) + 0.5;
+            }
+        }
+
+#           undef tween$bounceout
+
+        case TYPE::SINESQUARE: {
+            double A = sin((p)*pi2);
+            return A*A;
+        }
+
+        case TYPE::EXPONENTIAL: {
+            return 1 / (1 + exp(6 - 12 * (p)));
+        }
+
+        case TYPE::SCHUBRING1: {
+            return 2 * (p + (0.5f - p)*abs(0.5f - p)) - 0.5f;
+        }
+
+        case TYPE::SCHUBRING2: {
+            double p1pass = 2 * (p + (0.5f - p)*abs(0.5f - p)) - 0.5f;
+            double p2pass = 2 * (p1pass + (0.5f - p1pass)*abs(0.5f - p1pass)) - 0.5f;
+            double pAvg = (p1pass + p2pass) / 2;
+            return pAvg;
+        }
+
+        case TYPE::SCHUBRING3: {
+            double p1pass = 2 * (p + (0.5f - p)*abs(0.5f - p)) - 0.5f;
+            double p2pass = 2 * (p1pass + (0.5f - p1pass)*abs(0.5f - p1pass)) - 0.5f;
+            return p2pass;
+        }
+
+        case TYPE::SWING: {
+            return ((-cos(pi * p) * 0.5) + 0.5);
+        }
+
+        case TYPE::SINPI2: {
+            return sin(p * pi2);
+        }
+        }
+    }
+}
+
+namespace ImGui {
+    // [src] http://iquilezles.org/www/articles/minispline/minispline.htm
+    // key format (for dim == 1) is (t0,x0,t1,x1 ...)
+    // key format (for dim == 2) is (t0,x0,y0,t1,x1,y1 ...)
+    // key format (for dim == 3) is (t0,x0,y0,z0,t1,x1,y1,z1 ...)
+    void spline(const float *key, int num, int dim, float t, float *v) {
+        static signed char coefs[16] = {
+            -1, 2,-1, 0,
+            3,-5, 0, 2,
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') comma should be followed by whitespace [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AT009)